src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,275 @@
+#include "utils/builtins.h"
+#include "miscadmin.h"
+
+static bool IsRemoteReplicationOriginSessionSetup(MultiConnection *connection);
+
+static bool ExecuteRemoteCommandAndCheckResult(MultiConnection *connection, char *command,
+											   char *expected);
+
+static inline bool IsLocalReplicationOriginSessionActive(void);
+
+PG_FUNCTION_INFO_V1(replication_origin_session_start_no_publish);
+PG_FUNCTION_INFO_V1(replication_origin_session_end_no_publish);
+PG_FUNCTION_INFO_V1(replication_origin_session_is_no_publish);
+
+/*
+ * This variable is used to remember the replication origin id of the current session
+ * before resetting it to DoNotReplicateId in SetupReplicationOriginLocalSession.
+ */
+static RepOriginId originalOriginId = InvalidRepOriginId;
+
+/*
+ * Setting that controls whether replication origin sessions are enabled.
+ */
+bool isReplicationOriginSessionFeatureEnabled = false;
+
+
+/* replication_origin_session_start_no_publish starts a new replication origin session
+ * in the local node. This function is used to avoid publishing the WAL records to the
+ * replication slot by setting replication origin to DoNotReplicateId in WAL records.
+ * It remembers the previous replication origin for the current session which will be
+ * used to reset the replication origin to the previous value when the session ends.
+ */
+Datum
+replication_origin_session_start_no_publish(PG_FUNCTION_ARGS)
+{
+	if (!isReplicationOriginSessionFeatureEnabled)
+	{
+		PG_RETURN_VOID();
+	}
+	SetupReplicationOriginLocalSession();
+	PG_RETURN_VOID();
+}
+
+
+/* replication_origin_session_end_no_publish ends the current replication origin session
+ * in the local node. This function is used to reset the replication origin to the
+ * earlier value of replication origin.
+ */
+Datum
+replication_origin_session_end_no_publish(PG_FUNCTION_ARGS)
+{
+	if (!isReplicationOriginSessionFeatureEnabled)
+	{
+		PG_RETURN_VOID();
+	}
+	ResetReplicationOriginLocalSession();
+	PG_RETURN_VOID();
+}
+
+
+/* replication_origin_session_is_no_publish checks if the current replication origin
+ * session is active in the local node.
+ */
+Datum
+replication_origin_session_is_no_publish(PG_FUNCTION_ARGS)
+{
+	if (!isReplicationOriginSessionFeatureEnabled)
+	{
+		PG_RETURN_BOOL(false);
+	}
+	bool result = IsLocalReplicationOriginSessionActive();
+	PG_RETURN_BOOL(result);
+}
+
+
+/* IsLocalReplicationOriginSessionActive checks if the current replication origin
+ * session is active in the local node.
+ */
+inline bool
+IsLocalReplicationOriginSessionActive(void)
+{
+	return (replorigin_session_origin != InvalidRepOriginId);
+}
+
+
+/*
+ * SetupReplicationOriginLocalSession sets up a new replication origin session in a
+ * local session.
+ */
+void
+SetupReplicationOriginLocalSession(void)
+{
+	if (!isReplicationOriginSessionFeatureEnabled)
+	{
+		return;
+	}
+
+	/*elog(LOG, "Setting up local replication origin session"); */
+	if (!IsLocalReplicationOriginSessionActive())
```

Comment:
why do we have this check? Below 2 lines suggest that we want to preserve the original replication origin if one is already active

Reply:
I thought to check the feature flag and not doing anything new if this feature flag is not set..
So that no new code will get executed when the feature flag is turned off (by default).

Reply:
hmz, the definition of IsLocalReplicationOriginSessionActive is `replorigin_session_origin != InvalidRepOriginId` 

so we skip this logic if the replication_session_origin is set, but on the line below we explicitly try to save it (even though can only be InvalidRepOriginId?)

Reply:
Ok I have removed this check now.